### ISSUE TITLE
ShareIndex offset in `calculateCompactShareCount`

### ISSUE THREAD

**Issue by mpoke** [informalsystems/audit-celestia#25]
I'm not sure about the initialization of `msgSharesCursor` in the `calculateCompactShareCount` method, see https://github.com/celestiaorg/celestia-app/blob/e088d61fcb6579b4bc797deefd2ceff7601aa079/app/estimate_square_size.go#L72. This is used by `WrapMalleatedTx` to assign the `ShareIndex` for the first malleated tx. It's also used by `MsgSharesUsedNonInteractiveDefaults` to compute the number of shared used by the corresponding message. The number of shared used depends on the `cursor` input (i.e., `msgSharesCursor`), see https://github.com/celestiaorg/celestia-app/blob/e088d61fcb6579b4bc797deefd2ceff7601aa079/pkg/shares/non_interactive_defaults.go#L30.

**Comment by mpoke** [comment]
Now that I think about it, is this [line](https://github.com/celestiaorg/celestia-app/blob/e088d61fcb6579b4bc797deefd2ceff7601aa079/app/estimate_square_size.go#L80) even needed? The actual `ShareIndex`s are anyway recomputed [here](https://github.com/celestiaorg/celestia-app/blob/e088d61fcb6579b4bc797deefd2ceff7601aa079/app/malleate_txs.go#L64). Why not set the `ShareIndex` to zero for now?

**Comment by rootulp** [comment]
I removed the payment label because this issue seems related to app / shares packages but please re-add if I'm mistaken

**Comment by rootulp** [comment]
> I'm not sure about the initialization of msgSharesCursor in the calculateCompactShareCount method

Agreed. I would expect 

```diff
-	msgSharesCursor := len(txs)
+	msgSharesCursor := txSplitter.Count() + evdSplitter.Count()
```

**Comment by mpoke** [comment]
Yeah, but that count you can get only at the end. That's why you update the share indexes here https://github.com/celestiaorg/celestia-app/blob/e088d61fcb6579b4bc797deefd2ceff7601aa079/app/malleate_txs.go#L64

**Comment by rootulp** [comment]
> Why not set the `ShareIndex` to zero for now?

Are you proposing something like https://github.com/celestiaorg/celestia-app/pull/904 ? Sorry I need to investigate more to understand if this works*

**Comment by mpoke** [comment]
> > Why not set the `ShareIndex` to zero for now?
> 
> Are you proposing something like [celestiaorg/celestia-app#904](https://github.com/celestiaorg/celestia-app/pull/904) ? Sorry I need to investigate more to understand if this works*

Exactly
